### PR TITLE
linkers: Fix detection of link arguments to Clang(-cl) + MSVC

### PR DESCRIPTION
**Pull request by amyspark**
This PR fixes passing and detection of link arguments when using Microsoft's linker in combination with clang-cl or stock clang.

Currently, not only Meson lacks a way to induce a `--fatal-warnings` on LINK.exe, thus rendering all flag detection morose, it is also unable to pass flags appropriately by wrapping them in `-Wl,`.

This commit fixes it by implementing `fatal_warnings()` in the MSVCDynamicLinker and ClangCLDynamicLinker classes, and by implementing the requisite conversion steps in linker_to_compiler_args for ClangCompiler.

### CODE REVIEW

File: mesonbuild/compilers/mixins/clang.py [hunk 1]
```diff
@@ -165,6 +165,12 @@ def get_lto_compile_args(self, *, threads: int = 0, mode: str = 'default') -> T.
             args.extend(super().get_lto_compile_args(threads=threads))
         return args
 
+    def linker_to_compiler_args(self, args: T.List[str]) -> T.List[str]:
+        if isinstance(self.linker, (ClangClDynamicLinker, MSVCDynamicLinker)):
+            return list(map(lambda flag: flag if flag.startswith('-Wl,') else f'-Wl,{flag}', args))
```
Comment (nirbheek) [inline mesonbuild/compilers/mixins/clang.py:170]: Nitpick: this can just be a list comprehension:

```python
return [flag if flag.startswith('-Wl,') else f'-Wl,{flag}' for flag in args]
```

Much simpler to understand than list() + map() + lambda.

**Comment by nirbheek** [comment]
LGTM otherwise, but maybe @eli-schwartz or @xclaesse will want to comment since they're more up-to-date on the compilers/linkers classes.

File: mesonbuild/compilers/mixins/clang.py (hunk 1 above)
Comment (amyspark) [reply mesonbuild/compilers/mixins/clang.py:170]: Thanks!

File: mesonbuild/linkers/linkers.py [hunk 2]
```diff
@@ -1317,6 +1317,9 @@ def get_always_args(self) -> T.List[str]:
     def get_win_subsystem_args(self, value: str) -> T.List[str]:
         return self._apply_prefix([f'/SUBSYSTEM:{value.upper()}'])
 
+    def fatal_warnings(self) -> T.List[str]:
+        return ['-WX']
+
```
Comment (tristan957) [inline mesonbuild/linkers/linkers.py:1322]: Seems like these are unrelated changes? Might make sense to split them out into their own commit

Comment (amyspark) [reply mesonbuild/linkers/linkers.py:1322]: No, they're part of the same. See this project I'm testing it with: https://github.com/dragonCodecs/libusb/blob/blackmagic/meson/meson.build#L279

With the `-Wl` conversion, the `-WX` flag can make it past clang into the LINK.exe line. And `-WX` turns the warning

> unrecognized option '/-add-stdcall-alias'; ignored 

into a hard failure, which is what Meson uses for detecting unsupported flags.

For the specific error, see https://learn.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-warning-lnk4044?view=msvc-170

Comment (nirbheek) [reply mesonbuild/linkers/linkers.py:1322]: No? It's necessary afaict: https://learn.microsoft.com/en-us/cpp/build/reference/wx-treat-linker-warnings-as-errors?view=msvc-170

Comment (nirbheek) [reply mesonbuild/linkers/linkers.py:1322]: Sorry, my browser didn't load amy's comment.

**Comment by nirbheek** [comment]
@amyspark could you please add a test for this? I'm surprised this wasn't already caught by CI.

File: mesonbuild/compilers/mixins/clang.py [hunk 3]
```diff
@@ -165,6 +165,12 @@ def get_lto_compile_args(self, *, threads: int = 0, mode: str = 'default') -> T.
             args.extend(super().get_lto_compile_args(threads=threads))
         return args
 
+    def linker_to_compiler_args(self, args: T.List[str]) -> T.List[str]:
+        if isinstance(self.linker, (ClangClDynamicLinker, MSVCDynamicLinker)):
+            return [flag if flag.startswith('-Wl,') else f'-Wl,{flag}' for flag in args]
```
[the remark this answers is not in the document]
Comment (xclaesse) [reply mesonbuild/compilers/mixins/clang.py:170]: I think Meson has always assumed that linker arguments are already prefixed with `-Wl,`, see for example https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/meson.build#L86. That's probably because not all linker args can be prefixed that way? Surprisingly, this seems to not be even documented???? https://mesonbuild.com/Reference-manual_functions.html#add_project_link_arguments.

For consistency we should do it for all linkers or none... but I'm not sure why you do it only in this very specific case.

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply mesonbuild/compilers/mixins/clang.py:170]: > I think Meson has always assumed that linker arguments are already prefixed with `-Wl,`, see for example https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/main/subprojects/gstreamer/meson.build#L86. That's probably because not all linker args can be prefixed that way?

Indeed, the meaning of `-Wl,` means that the compiler driver should not try to interpret it, but rather pass it unmodified to the linker -- which is very different from "flags that should be used when linking, not when compiling".

Examples where this fails would include `-fuse-ld=bfd`, which the compiler driver must interpret to decide which linker to execute and which linkers don't understand, or `-static` which acts differently from `-Wl,-static` as the former tells the compiler driver to compile a static binary, while the latter tells the linker alone to just load `*.a` archives and may be used to push/pop some static libraries and some shared ones.

[the remark this answers is not in the document]
Comment (amyspark) [reply mesonbuild/compilers/mixins/clang.py:170]: In the combination of Clang + LINK.EXE, assuming all flags are wrapped with `-Wl` is *not* correct; all linker flags, *including the ones internally supplied by Meson*, must be forwarded with `-Wl`.

```
-----------
Command line: `clang C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmp1xjsk_eb\testfile.c -o C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmp1xjsk_eb\output.exe -D_FILE_OFFSET_BITS=64 -O0 -Werror=implicit-function-declaration -WX -Wl,--add-stdcall-alias` -> 0
stdout:
LINK : warning LNK4044: unrecognized option '/-add-stdcall-alias'; ignored
-----------
stderr:
warning: unknown warning option '-WX' [-Wunknown-warning-option]
1 warning generated.
-----------
Compiler for C supports link arguments -Wl,--add-stdcall-alias: YES 
```

I did not notice, however, that in clang-cl we've got double nesting (and a broken linker detection-- see who it's calling for the linking process):

```
❯ py -3 E:\meson\meson.py setup build --wipe 
The Meson build system
Version: 1.2.99
Source dir: C:\Users\Amalia\Desktop\Rachel\bmpflash
Build dir: C:\Users\Amalia\Desktop\Rachel\bmpflash\build
Build type: native build
Project name: bmpflash
Project version: 0.0.1
Activating VS 17.8.0 Preview 1.0
C++ compiler for the host machine: clang-cl (clang-cl 16.0.5)
C++ linker for the host machine: lld-link lld-link 16.0.5
```

```
-----------
Command line: `clang-cl C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpwu6gdz3a\testfile.cpp /FeC:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpwu6gdz3a\output.exe /nologo /showIncludes /utf-8 /MD /nologo /showIncludes /utf-8 /Od /Oi- /link -WX -guard:cf` -> 0
Compiler for C++ supports link arguments -guard:cf: YES 
Running compile:
Working directory:  C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4
Code:
 int main(void) { return 0; }

-----------
Command line: `clang-cl C:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4\testfile.cpp /FeC:\Users\Amalia\Desktop\Rachel\bmpflash\build\meson-private\tmpt3_2ydx4\output.exe /nologo /showIncludes /utf-8 /MD /nologo /showIncludes /utf-8 /Od /Oi- /link -WX -Wl,-guard:cf` -> 4044
stdout:
LINK : warning LNK4044: unrecognized option '/Wl,-guard:cf'; ignored
LINK : error LNK1218: warning treated as error; no output file generated
-----------
stderr:
clang-cl: error: linker command failed with exit code 4044 (use -v to see invocation)
-----------
Compiler for C++ supports link arguments -Wl,-guard:cf: NO
```

[the remark this answers is not in the document]
Comment (amyspark) [reply mesonbuild/compilers/mixins/clang.py:170]: Sent a test to cover both cases, and fixed the linker preference forwarding for clang-cl.

File: mesonbuild/compilers/mixins/visualstudio.py [hunk 4]
```diff
@@ -489,6 +489,18 @@ def get_include_args(self, path: str, is_system: bool) -> T.List[str]:
...
+    def use_linker_args(cls, linker: str, version: str) -> T.List[str]:
+        # Clang additionally can use a linker specified as a path, unlike MSVC.
+        if linker == 'lld-link':
+            return ['-fuse-ld=lld-link']
+        return super().use_linker_args(linker, version)
+
+    def linker_to_compiler_args(self, args: T.List[str]) -> T.List[str]:
+        # clang-cl forwards arguments span-wise with the /LINK flag
+        # therefore -Wl will be received by lld-link or LINK and rejected
+        return self.__class__.use_linker_args(self.linker.id, '') + super().linker_to_compiler_args([flag[4:] if flag.startswith('-Wl,') else flag for flag in args])
```
[the remark this answers is not in the document]
Comment (amyspark) [reply mesonbuild/compilers/mixins/visualstudio.py:502]: flake8 does not recognise class methods:
```
mesonbuild/compilers\mixins\visualstudio.py:502:16: F821 undefined name 'use_linker_args'
```

[the remark this answers is not in the document]
Comment (amyspark) [reply mesonbuild/compilers/mixins/visualstudio.py:502]: Removing in favour of `super()`.

**Comment by tristan957** [comment]
@bruchar1 can you test this?
